pkg/services/vmoperator/vmopmachine.go

```diff
@@ -188,6 +193,116 @@ func (v *VmopMachineService) ReconcileNormal(ctx context.Context, machineCtx cap
 	// Set the VM state. Will get reset throughout the reconcile
 	supervisorMachineCtx.VSphereMachine.Status.VMStatus = vmwarev1.VirtualMachineStatePending
 
+	var affInfo affinityInfo
+	if feature.Gates.Enabled(feature.NodeAutoPlacement) &&
+		!infrautilv1.IsControlPlaneMachine(machineCtx.GetVSphereMachine()) {
+		vmOperatorVMGroup := &vmoprv1.VirtualMachineGroup{}
+		key := client.ObjectKey{
+			Namespace: supervisorMachineCtx.Cluster.Namespace,
+			Name:      supervisorMachineCtx.Cluster.Name,
+		}
+		err := v.Client.Get(ctx, key, vmOperatorVMGroup)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return false, err
+			}
+			if apierrors.IsNotFound(err) {
+				log.V(4).Info("VirtualMachineGroup not found, requeueing", "Name", key.Name, "Namespace", key.Namespace)
+				return true, nil
+			}
+		}
+
+		// Proceed only if the machine is a member of the VirtualMachineGroup.
+		isMember, err := v.checkVirtualMachineGroupMembership(vmOperatorVMGroup, supervisorMachineCtx)
+		if err != nil {
+			return true, errors.Wrapf(err, "%s", fmt.Sprintf("failed to check if VirtualMachine %s is a member of VirtualMachineGroup %s/%s", supervisorMachineCtx.VSphereMachine.Name, vmOperatorVMGroup.Name, vmOperatorVMGroup.Namespace))
+		}
+		if !isMember {
+			v1beta2conditions.Set(supervisorMachineCtx.VSphereMachine, metav1.Condition{
+				Type:   infrav1.VSphereMachineVirtualMachineProvisionedV1Beta2Condition,
+				Status: metav1.ConditionFalse,
+				Reason: infrav1.VSphereMachineVirtualMachineWaitingForVirtualMachineGroupV1Beta2Reason,
+			})
+			log.V(4).Info("Waiting for VirtualMachineGroup membership, requeueing", "VM Name", supervisorMachineCtx.Machine.Name)
+			return true, nil
+		}
+
+		affInfo = affinityInfo{
+			vmGroupName: vmOperatorVMGroup.Name,
+		}
+
+		// Set the zone label using the annotation of the per-md zone mapping from VMG.
+		// This is for new VMs created during day-2 operations in VC 9.1.
+		nodePool := supervisorMachineCtx.Machine.Labels[clusterv1.MachineDeploymentNameLabel]
+		if zone, ok := vmOperatorVMGroup.Annotations[fmt.Sprintf("zone.cluster.x-k8s.io/%s", nodePool)]; ok && zone != "" {
+			affInfo.failureDomain = ptr.To(zone)
+		}
+
+		// Fetch machine deployments without explicit failureDomain specified
```

Comment:
I'm wondering if we should skip computing AffinitySpec after the machine has been created (we not using it for update)

Reply:
AffinitySpec doesn't impact after the initial cluster creation, but to keep it consistent, I think we should keep computing here. Especially in future, if AffinitySpec start to take effect during day-2, we don't need to update the code here.